src/k2/module/k23si/TxnWIMetaManager.cpp

```diff
@@ -0,0 +1,461 @@
+            twim.trh,
+            twim.mtr.timestamp});
+
+    auto fut = _cpo.partitionRequest<dto::K23SIInspectTxnRequest, dto::K23SIInspectTxnResponse, dto::Verbs::K23SI_INSPECT_TXN>(Deadline(_config.readTimeout()), *request)
+        .then([this, &twim, request=std::move(request)] (auto&& result) {
+            (void)request; // to be deleted by unique_ptr
+            auto& [status, resp] = result;
+            K2LOG_D(log::skvsvr, "inspect for twim {}, received response status={}, resp={}", twim, status, resp);
+            // only trigger action if the TRH gives a response. Otherwise we have to retry
+            // for all 3 cases below, we just rely on the RWE timer to trigger the next transition
+            twim.unlinkRW(_rwlist);
+            _rwlist.push_back(twim);
+            if (status.is2xxOK()) {
+                return seastar::make_ready_future<Status>(_onAction(resp.finalizeAction == dto::EndAction::Commit ? Action::onCommit : Action::onAbort, twim));
+            }
+            else if (status == dto::K23SIStatus::KeyNotFound) {
+                return seastar::make_ready_future<Status>(_onAction(Action::onAbort, twim));
+            }
+            K2LOG_W(log::skvsvr, "Unable to communicate with TRH for {} after RWE", twim);
+            return seastar::make_ready_future<Status>(status);
+        });
+    _addBgTask(twim, [fut=std::move(fut)]() mutable { return fut.discard_result(); });
+    return Statuses::S200_OK("processed state transition");
+}
+
+Status TxnWIMetaManager::_finalizing(TxnWIMeta& twim) {
+    auto newState = dto::TxnWIMetaState::Finalizing;
+    K2LOG_D(log::skvsvr, "Entering state: {}", newState);
+    twim.state = newState;
+    twim.unlinkRW(_rwlist);
+    return Statuses::S200_OK("processed state transition");
+}
+
+Status TxnWIMetaManager::_finalizedPIP(TxnWIMeta& twim) {
+    auto newState = dto::TxnWIMetaState::FinalizedPIP;
+    K2LOG_D(log::skvsvr, "Entering state: {}", newState);
+    twim.state = newState;
+
+    auto fut = _persistence->append_cont(twim)
+        .then([this, &twim] (auto&& status) {
+            K2LOG_D(log::skvsvr, "persist completed for {} with {}", twim, status);
+            if (!status.is2xxOK()) {
+                // flush didn't succeed
+                K2LOG_E(log::skvsvr, "persist failed for {} with {}", twim, status);
+                return seastar::make_ready_future<Status>(_onAction(Action::onPersistFail, twim));
+            }
+            return seastar::make_ready_future<Status>(_onAction(Action::onPersistSucceed, twim));
+        });
+    _addBgTask(twim, [fut=std::move(fut)] () mutable { return fut.discard_result();});
+    return Statuses::S201_Created("Finalized twim");
+}
+
+Status TxnWIMetaManager::_finalized(TxnWIMeta& twim, bool success) {
+    // we don't have an actual state to model this - we just remove the record from tracking
+    K2LOG_D(log::skvsvr, "Entering state: Finalized");
+    _twims.erase(twim.mtr.timestamp);
```

Comment:
should we assert twim doesn't have any written keys here?

Reply:
done